File: Tests/AINotebookCoreTests/NotebookStoreSourcesTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class NotebookStoreSourcesTests: XCTestCase {
    func testCreateAndFetchSources() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")

        let s1 = try store.createSource(
            notebookId: nb.id!,
            type: .text,
            title: "Alpha",
            uri: nil,
            rawPath: nil
        )
        let s2 = try store.createSource(
            notebookId: nb.id!,
            type: .pdf,
            title: "Beta",
            uri: nil,
            rawPath: "/tmp/beta.pdf"
        )

        let list = try store.sources(notebookId: nb.id!)
        XCTAssertEqual(list.count, 2)
        XCTAssertEqual(Set(list.map(\.title)), ["Alpha", "Beta"])
        XCTAssertEqual(s1.status, .pending)
        XCTAssertEqual(s2.status, .pending)
    }

    func testUpdateStatusPersists() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "X", uri: nil, rawPath: nil
        )
        try store.updateSourceStatus(id: s.id!, status: .ready, error: nil)
        let reloaded = try XCTUnwrap(store.source(id: s.id!))
        XCTAssertEqual(reloaded.status, .ready)
        XCTAssertNil(reloaded.error)
    }

    func testUpdateStatusErrorPersists() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "X", uri: nil, rawPath: nil
        )
        try store.updateSourceStatus(id: s.id!, status: .error, error: "boom")
        let reloaded = try XCTUnwrap(store.source(id: s.id!))
        XCTAssertEqual(reloaded.status, .error)
        XCTAssertEqual(reloaded.error, "boom")
    }

    func testReplaceChunksClearsPreviousAndInserts() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "X", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [
                ChunkDraft(text: "one",   tokenCount: 1, pageHint: nil),
                ChunkDraft(text: "two",   tokenCount: 1, pageHint: nil),
                ChunkDraft(text: "three", tokenCount: 1, pageHint: nil)
            ]
        )
        let first = try store.chunks(sourceId: s.id!)
        XCTAssertEqual(first.count, 3)
        XCTAssertEqual(first.map(\.ord), [0, 1, 2])
        XCTAssertEqual(first.map(\.text), ["one", "two", "three"])

        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [ChunkDraft(text: "only", tokenCount: 1, pageHint: nil)]
        )
        let second = try store.chunks(sourceId: s.id!)
        XCTAssertEqual(second.count, 1)
        XCTAssertEqual(second[0].text, "only")
    }

    func testDeleteSourceCascadesChunks() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "X", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [ChunkDraft(text: "x", tokenCount: 1, pageHint: nil)]
        )
        try store.deleteSource(id: s.id!)
        XCTAssertNil(try store.source(id: s.id!))
        XCTAssertEqual(try store.chunks(sourceId: s.id!).count, 0)
    }

    func testCreateRejectsEmptyTitle() {
        do {
            let store = try NotebookStore(path: .inMemory)
            let nb = try store.createNotebook(name: "NB")
            _ = try store.createSource(
                notebookId: nb.id!, type: .text, title: "   ", uri: nil, rawPath: nil
            )
            XCTFail("expected throw")
        } catch StoreError.invalidSourceTitle {
            // ok
        } catch {
            XCTFail("wrong error: \(error)")
        }
    }

    func testSourcesExcludesNoteShadowRows() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        _ = try store.createSource(
            notebookId: nb.id!, type: .text, title: "Real",
            uri: nil, rawPath: nil
        )
        _ = try store.createSource(
            notebookId: nb.id!, type: .note, title: "Shadow",
            uri: nil, rawPath: nil
        )
        let visible = try store.sources(notebookId: nb.id!)
        XCTAssertEqual(visible.count, 1)
        XCTAssertEqual(visible.first?.title, "Real")

        let all = try store.sourcesIncludingShadow(notebookId: nb.id!)
        XCTAssertEqual(all.count, 2)
    }
}

File: windows/tests/AINotebook.Core.Tests/Storage/NotebookStoreSourcesTests.cs
using AINotebook.Core;
using AINotebook.Core.Models;
using AINotebook.Core.Storage;
using Xunit;

namespace AINotebook.Core.Tests.Storage;

public class NotebookStoreSourcesTests
{
    private static (NotebookStore store, long nbId) Fresh()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N");
        return (store, nb.Id!.Value);
    }

    [Fact]
    public void CreateSourceDefaultsToPendingAndTrimsTitle()
    {
        var (store, nb) = Fresh();
        using (store)
        {
            var s = store.CreateSource(nb, SourceType.Text, "  Doc  ", uri: null, rawPath: "/tmp/x");
            Assert.Equal(SourceStatus.Pending, s.Status);
            Assert.Equal("Doc", s.Title);
        }
    }

    [Fact]
    public void CreateSourceRejectsEmptyTitle()
    {
        var (store, nb) = Fresh();
        using (store)
            Assert.Throws<StoreException.InvalidSourceTitle>(
                () => store.CreateSource(nb, SourceType.Text, "  ", null, null));
    }

    [Fact]
    public void UpdateSourceStatusPersistsAndUnknownThrows()
    {
        var (store, nb) = Fresh();
        using (store)
        {
            var s = store.CreateSource(nb, SourceType.Pdf, "p", null, null);
            store.UpdateSourceStatus(s.Id!.Value, SourceStatus.Error, "boom");
            var reloaded = store.Source(s.Id!.Value)!;
            Assert.Equal(SourceStatus.Error, reloaded.Status);
            Assert.Equal("boom", reloaded.Error);
            Assert.Throws<StoreException.SourceNotFound>(
                () => store.UpdateSourceStatus(99999, SourceStatus.Ready, null));
        }
    }

    [Fact]
    public void UpdateSourceTitleChangesTitleAndSyncsFts()
    {
        var (store, nb) = Fresh();
        using (store)
        {
            var s = store.CreateSource(nb, SourceType.Text, "OldTitle", null, null);
            store.UpdateSourceTitle(s.Id!.Value, "BrandNewTitle");
            Assert.Equal("BrandNewTitle", store.Source(s.Id!.Value)!.Title);

            // The sources_au trigger must keep sources_fts in sync with the new title.
            using var cmd = store.Connection.CreateCommand();
            cmd.CommandText = "SELECT count(*) FROM sources_fts WHERE sources_fts MATCH $q";
            cmd.Parameters.AddWithValue("$q", "BrandNewTitle");
            Assert.Equal(1L, (long)cmd.ExecuteScalar()!);
        }
    }

    [Fact]
    public void ReplaceChunksClearsThenReinsertsWithOrdZeroToN()
    {
        var (store, nb) = Fresh();
        using (store)
        {
            var s = store.CreateSource(nb, SourceType.Text, "t", null, null);
            store.ReplaceChunks(s.Id!.Value, new[]
            {
                new ChunkDraft("first", 1, null),
                new ChunkDraft("second", 1, 2),
            });
            // Replace again with a single chunk -> old ones cleared
            store.ReplaceChunks(s.Id!.Value, new[] { new ChunkDraft("only", 1, null) });
            var chunks = store.Chunks(s.Id!.Value);
            Assert.Single(chunks);
            Assert.Equal(0, chunks[0].Ord);
            Assert.Equal("only", chunks[0].Text);
        }
    }

    [Fact]
    public void DeleteSourceCascadesChunks()
    {
        var (store, nb) = Fresh();
        using (store)
        {
            var s = store.CreateSource(nb, SourceType.Text, "t", null, null);
            store.ReplaceChunks(s.Id!.Value, new[] { new ChunkDraft("a", 1, null) });
            store.DeleteSource(s.Id!.Value);
            Assert.Empty(store.Chunks(s.Id!.Value));
            Assert.Throws<StoreException.SourceNotFound>(() => store.DeleteSource(s.Id!.Value));
        }
    }

    [Fact]
    public void SourceSummaryDefaultsToNullThenRoundTrips()
    {
        var (store, nb) = Fresh();
        using (store)
        {
            var s = store.CreateSource(nb, SourceType.Text, "t", null, null);
            Assert.Null(store.SourceSummary(s.Id!.Value));
            store.SetSourceSummary(s.Id!.Value, "A short summary.");
            Assert.Equal("A short summary.", store.SourceSummary(s.Id!.Value));
        }
    }

    [Fact]
    public void SourcesExcludesShadowNotesButIncludingShadowReturnsThem()
    {
        var (store, nb) = Fresh();
        using (store)
        {
            store.CreateSource(nb, SourceType.Text, "real", null, null);
            store.CreateSource(nb, SourceType.Note, "shadow", null, null);
            Assert.Single(store.Sources(nb));
            Assert.Equal("real", store.Sources(nb).Single().Title);
            Assert.Equal(2, store.SourcesIncludingShadow(nb).Count);
        }
    }
}
